### DISCUSSION TITLE
How to lazyload nvim-osc52?

### DISCUSSION THREAD

**Discussion by limaceous-bushwhacker** [LazyVim/LazyVim#1883] [Q&A] [answered]
https://github.com/ojroques/nvim-osc52

**Comment by dpetka2001** [comment]
Just take a look at [yanky](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/coding/yanky.lua) configuration in Lazyvim. If you use the `keys` handlers to invoke the plugins' actions, then it will be lazy loaded.

**Reply by limaceous-bushwhacker** [reply]
Appreciate the quick response. I've managed to get it to work in visual mode but the bindings in normal mode doesn't seem to work. Here's what I'm doing in the keys section: `{ "cp", function() require ("osc52").copy_operator end, { expr = true }, map = "n" }`. This show allow me to do `cpiw` but it's not working as expected. If I don't lazyload using `keys` it seems to work fine.

**Reply by dpetka2001** [reply, marked as answer]
```lua
      {
        "<leader>cp",
        function()
          return require("osc52").copy_operator()
        end,
        expr = true,
        desc = "osc52",
      },
```
seems to do the trick after some experimenting.

**Reply by limaceous-bushwhacker** [reply]
Thank you so much mate. Finally got it to work. The `expr = true` section had me scratching my head for hours. Looks like I need to pick up some Lua.

**Reply by ninbura** [reply]
Was just stuck on this same issue for the past 3 days. 

Search engines weren't linking to this thread. Was about to create my own issue, but did a search directly in LazyVim's discussion section before posting. My problem was that I wasn't using a `return` statement on the `copy_operator` call. What's weird is that I don't have to use a return statement on the `copy_visual` call.
```lua
return {
  "ojroques/nvim-osc52",
  keys = {
    {
      "<leader>y",
      function() return require("osc52").copy_operator() end,
      desc = "copy selection to system clipboard (normal mode)",
      expr = true
    },
    {
      "<leader>Y",
      "<leader>y_",
      remap = true,
      desc = "copy current line into system clipboard (normal mode)"
    },
    {
      mode = "v",
      "<leader>y",
      function() require("osc52").copy_visual() end,
      desc = "copy selection to system clipboard (visual mode)"
    }
  },
  config = function()
    -- vim.keymap.set("n", "<leader>y", require("osc52").copy_operator, { expr = true })
  end
}
```

I eventually deduced it had something to do with the `expr` property after messing with the keymap using native lua syntax, as seen in the comment at the bottom of the block. But I still don't actually understand what's going on there...

Why does that function need a return statement? What is `expr` in this context?

**Reply by dpetka2001** [reply]
That depends on what each function returns. 

`copy_visual` just returns `M.copy(text)`, which in turn also return something specific that doesn't need evaluating. 

`copy_operator` function returns a call to the function that is set by the `operatorfunc` option, thus needs to be evaluated to actually return the `rhs` part that `vim.keymap.set()` expects as an argument. 

`expr = true` is set when the `rhs` isn't directly what is types, but instead has to be evaluated by an expression to produce the `rhs`.
